File: Testing/Python/TreeNode.py

class BinaryTree:
    def __init__(self, list = None):
        self.root = None
        if list is not None:
            self.__create_tree_from_list(list)

    
                        
    def __create_tree_from_list(self, list_of_nodes):
        '''
        Creates a BinaryTree from the list of nodes in level order.
        Note: you must specify *every* node on a given level, even if its nil.
        
        '''
        if len(list_of_nodes) == 0:
            return 

        #if the root is null, you can't reach the rest of the nodes
        if list_of_nodes[0] is not None:
            self.root = TreeNode(list_of_nodes[0])
        if list_of_nodes[0] is None or len(list_of_nodes) == 1:
            return
                
        cur_level = []
        next_level = []
        cur_level.append(self.root)
        i = 1
        while cur_level:
            cur = cur_level.pop(0)
            

            if cur is None:
                for x in range(2):
                    next_level.append(None)
                    i += 1
            else:    
                if list_of_nodes[i] is not None:
                    cur.left = TreeNode(list_of_nodes[i])
                else:
                    cur.left = None
                next_level.append(cur.left)
                i += 1
                if i == len(list_of_nodes):
                    return

                if list_of_nodes[i] is not None:
                    cur.right = TreeNode(list_of_nodes[i])
                else:
                    cur.right = None
                next_level.append(cur.right)
                i += 1
                if i == len(list_of_nodes):
                    return

            if not cur_level:
                cur_level = next_level
                next_level = []
            
    def __str__(self):
        '''
        Returns a level-order representation of the tree
        '''
        string = []
        cur_level = []
        next_level = []
        should_continue = False
        cur_level.append(self.root)
        string.append(str(self.root))
        string.append('\n')
        while True:
            cur = cur_level.pop(0)
            
            if cur is None:
                for i in range(2):
                    next_level.append(None)
                    string.append(str(None))
            else:
                #check left node
                if cur.left is not None:
                    should_continue = True
                
                next_level.append(cur.left)
                string.append(str(cur.left))

                #check right node
                if cur.right is not None:
                    should_continue = True

                next_level.append(cur.right)
                string.append(str(cur.right))
                    
                    
            #No more nodes            
            if not should_continue and not cur_level:
                break

            if not cur_level:
                string.append('\n')
                cur_level = next_level
                next_level = []
                should_continue = False
        return ''.join(string)

class TreeNode:
    def __init__(self, val):
        self.val = val
        self.right = None
        self.left = None
        
    def __str__(self):
        return '{self.val}'.format(self=self)

if __name__ == '__main__':
    test = [0,1,2,3,4,5,6,7,8]
    my_tree = BinaryTree(test)
    print "my Tree"
    print my_tree
    test2 = [0, None, 1, None, None, 2, 3, None, None, None, None, None, 5]
    my_tree2 = BinaryTree(test2)
    print "my tree2"
    print my_tree2

File: Testing/Java/TreeNode.java
package Testing.Java;

import java.util.*;

public class TreeNode{

    public int val;
    public TreeNode right;
    public TreeNode left;

    public TreeNode(int val){
	this.val = val;
	right = null;
	left = null;
    }

  public String toString(){
      return val + "";
  }

    /*
     Creates a tree from the Integer array and returns the root of the tree.
     */
    public static TreeNode createTreeFromArray(Integer[] array){
	if(array == null || array.length == 0){
	    return null;
	}

	TreeNode root = new TreeNode(array[0]);
	if(array.length == 1){
	    return root;
	}
	Deque<TreeNode> oldStack = new LinkedList<>();
	Deque<TreeNode> newStack = new LinkedList<>();
	oldStack.push(root);

	int i = 1;
	while(true){
	    TreeNode curNode = oldStack.pollLast();
	    System.out.println(curNode);
	    
	    TreeNode left = null;
	    TreeNode right= null;
	    if(curNode != null){
		//get our left node
		if(array[i] != null){
		    left = new TreeNode(array[i]);
		}
		
		
		//get our right node
		if(i + 1 < array.length && array[i + 1] != null){
		    right = new TreeNode(array[i + 1]);
		}
		
		curNode.right = right;
		curNode.left = left;
	    }
	    i += 2;
	    
	    //  System.out.println(left);
	    //System.out.println(right);
	    	   
	    newStack.push(left);
	    newStack.push(right);
	   

	    if(i >= array.length) break;
	    else if(oldStack.isEmpty()){
		//if we're done with this level
		oldStack = newStack;
		newStack = new LinkedList<>();
	    }
	}

	return root;
    }


    public static String getTreeAsString(TreeNode root){
	if(root == null) return "";
	
	StringBuilder str = new StringBuilder();
	
	Deque<TreeNode> oldStack = new LinkedList<>();
	Deque<TreeNode> newStack = new LinkedList<>();
	oldStack.push(root);
	str.append(root + "\n");
	boolean shouldContinue = false;

	while(true){
	    TreeNode cur = oldStack.pollLast();

	    TreeNode right = null;
	    TreeNode left = null;

	    if(cur == null){
		str.append("##");
	    }else{
		
		if(cur.left != null){
		    str.append(cur.left.val);
		    shouldContinue = true;
		    left = cur.left;
		}else{
		    str.append("#");
		}

		//	System.out.println(cur.right);

		if(cur.right != null){
		    str.append(cur.right.val);
		    shouldContinue = true;
		    right = cur.right;
		}else{
		    str.append("#");
		}
	    }
	    
	    newStack.push(left);
	    newStack.push(right);
	    if(oldStack.isEmpty()){
		if(!shouldContinue) break;
		str.append("\n");
		shouldContinue = false;
		oldStack = newStack;
		newStack = new LinkedList<>();
	    }
	}
	
	
	return str.toString();
    }

    public static void main(String[] args){
	TreeNode tree = createTreeFromArray(new Integer[]{1,2,3,4,5,6});
	System.out.println(getTreeAsString(tree));
	
    }

}
